src/Components/WebAssembly/WebAssembly/src/Hosting/MutableConfiguration.cs

```diff
@@ -0,0 +1,197 @@
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Hosting
+{
+    /// <summary>
+    /// MutableConfiguration is a class that implements the interface of an IConfiguration,
+    /// IConfigurationRoot, and IConfigurationBuilder. It can be used to simulatneously build
+    /// and read from a configuration object.
+    /// </summary>
+    public class MutableConfiguration : IConfiguration, IConfigurationRoot, IConfigurationBuilder
+    {
+        private readonly IList<IConfigurationProvider> _providers = new List<IConfigurationProvider>();
+        private readonly IList<IConfigurationSource> _sources = new List<IConfigurationSource>();
+
+        private readonly IList<IDisposable> _changeTokenRegistrations = new List<IDisposable>();
+        private ConfigurationReloadToken _changeToken = new ConfigurationReloadToken();
+
+        // Lock for setters and getters on the configuration object.
+        private readonly object mutex = new object();
+
+        /// <summary>
+        /// Returns the sources used to obtain configuration values.
+        /// </summary>
+        public IList<IConfigurationSource> Sources => _sources;
+
+        /// <summary>
+        /// Returns the providers user to obtain configuration values.
+        /// </summary>
+        public IEnumerable<IConfigurationProvider> Providers => _providers;
+
+        /// <summary>
+        /// Gets a key/value collection that can be used to share data between the <see cref="IConfigurationBuilder"/>
+        /// and the registered <see cref="IConfigurationProvider"/>s.
+        ///
+        /// In this implementation, this largely exists as a way to satisfy the
+        /// requirements of the IConfigurationBuilder and is not populated by
+        /// <see cref="MutableConfiguration"/> with any meaningful info.
+        /// </summary>
+        public IDictionary<string, object> Properties { get; } = new Dictionary<string, object>();
+
+        /// <summary>
+        /// Gets or sets a configuration value in all providers that are currently
+        /// registered on the <see cref="MutableConfiguration"/> object.
+        /// </summary>
+        /// <param name="key">The configuration key.</param>
+        /// <returns>The configuration value.</returns>
+        public string this[string key] {
+            get {
+                lock (mutex)
+                {
+                    for (var i = _providers.Count - 1; i >= 0; i--)
+                    {
+                        var provider = _providers[i];
+
+                        if (provider.TryGet(key, out var value))
+                        {
+                            return value;
+                        }
+                    }
+                }
+                return null;
+            }
+            set {
```

Comment:
Do we need the setter here?

I guess I imagined that users would use this object to:
- Read keys (based on current/initial state)
- Add additional providers

Reply:
I did imagine that users would be able to set values here. This configuration object is never frozen. IMO, this enables a lot of interesting scenarios for users so I am partial to keeping it.

Reply:
The reason I bring this up is that it adds some subtlety to the model that that wouldn't be there without it.

FYI `Set(...)` is generally implemented in a non-threadsafe way across all providers: https://github.com/dotnet/runtime/blob/master/src/libraries/Microsoft.Extensions.Configuration/src/ConfigurationProvider.cs#L25

Dictionary is thread-safe for multiple readers, as long as no one is writing. 

--- 

I actually wasn't even aware that `Set(...)` existed until your PR.